network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,390 @@
+
+use futures::io::{AsyncRead, AsyncReadExt, AsyncWrite, AsyncWriteExt};
+use std::{
+    collections::HashMap,
+    io,
+    sync::{Arc, RwLock},
+    time,
+};
+
+use libra_config::config::NetworkPeerInfo;
+use libra_crypto::{noise, x25519};
+use libra_types::PeerId;
+use netcore::{
+    negotiate::{negotiate_inbound, negotiate_outbound_interactive},
+    transport::ConnectionOrigin,
+};
+
+use crate::noise_wrapper::session::NoiseSession;
+
+// Timestamp
+// --------
+// We parse the client message and then enforce two things:
+// - the client sent us a timestamp in a range [-60s, +10s]
+// - the client is not re-using an older timestamp from that range
+//
+// if these are invalid, we avoid computing crypto to perform the answer
+// if these are valid, we store the new timestamp (even if it's older than the last received)
+// the last point assumes that a real client would not do this.
+// this is not a measure against real validators spamming us
+
+/// we're willing to tolerate a client timestamp 10 seconds in the future
+const MAX_FUTURE_TIMESTAMP: u64 = 10;
+
+/// we store client timestamps (to prevent replay) for up to 60 seconds
+const EXPIRATION_TIMESTAMP: u64 = 60;
+
+/// hashmap to store client timestamps if a connection succeeds
+pub type AntiReplayTimestamp = HashMap<x25519::PublicKey, u64>;
+
+// Noise Wrapper
+// -------------
+// Noise by default is not aware of the above or lower protocol layers,
+// We thus need to build this wrapper around Noise to both:
+//
+// - fragment messages that need to be encrypted by noise (due to its maximum 65535-byte messages)
+// - understand how long noise messages we send and receive are,
+//   in order to pass them to the noise implementaiton
+//
+
+/// TODO: why do we expose this to the outer protocol?
+const NOISE_PROTOCOL: &[u8] = b"/noise_ik_25519_aesgcm_sha256/1.0.0";
+
+/// The Noise configuration to be used to perform a protocol upgrade on an underlying socket.
+pub struct NoiseWrapper(noise::NoiseConfig);
+
+impl NoiseWrapper {
+    /// Create a new NoiseConfig with the provided keypair
+    pub fn new(key: x25519::PrivateKey) -> Self {
+        Self(noise::NoiseConfig::new(key))
+    }
+
+    /// Perform a protocol upgrade on an underlying connection. In addition perform the noise IX
+    /// handshake to establish a noise session and exchange static public keys. Upon success,
+    /// returns the static public key of the remote as well as a NoiseSession.
+    // TODO(mimoo, philp9): this code could be inlined in transport.rs once the monolithic network is done
+    pub async fn upgrade_connection<TSocket>(
+        &self,
+        socket: TSocket,
+        origin: ConnectionOrigin,
+        anti_replay_timestamps: Arc<RwLock<AntiReplayTimestamp>>,
+        remote_public_key: Option<x25519::PublicKey>,
+        trusted_peers: Option<&Arc<RwLock<HashMap<PeerId, NetworkPeerInfo>>>>,
+    ) -> io::Result<(x25519::PublicKey, NoiseSession<TSocket>)>
```

Comment:
since the protocol is not symmetric, we should have two separate functions

`upgrade_inbound` for when you're a listener
`upgrade_outbound` for when you're a dialer

the args will also change (e.g., we don't need origin, we don't need replay timestamps if dialer, we don't need remote_public_key if listener)